ocaml/typing/jkind.ml

```diff
@@ -500,31 +549,66 @@ let of_const ~why : const -> t = function
   | Void -> fresh_jkind (Sort Sort.void) ~why
   | Float64 -> fresh_jkind (Sort Sort.float64) ~why
 
-(* CR layouts v1.5: remove legacy_immediate *)
-let of_annotation ?(legacy_immediate = false) ~context
-    Location.{ loc; txt = const } =
-  (match const with
-  | (Immediate | Immediate64 | Value) when legacy_immediate -> ()
-  | _ ->
+let check_user_written_annotation ?(legacy_immediate = false) ~context ~loc
+    annot =
+  match annot with
+  | (Immediate | Immediate64 | Value) as const when legacy_immediate -> const
+  | const ->
     let required_layouts_level = get_required_layouts_level context const in
     if not (Language_extension.is_at_least Layouts required_layouts_level)
-    then raise ~loc (Insufficient_level (context, const)));
-  of_const ~why:(Annotated (context, loc)) const
+    then raise ~loc (Insufficient_level (context, const));
+    const
+
+let const_of_user_written_annotation ?legacy_immediate ~context
+    Location.{ loc; txt = annot } =
+  match Const.of_user_written_annotation_unchecked annot with
+  | None -> raise ~loc (Unknown_jkind annot)
+  | Some unchecked ->
+    check_user_written_annotation ?legacy_immediate ~context ~loc unchecked
+
+let const_to_user_written_annotation = Const.to_user_written_annotation
+
+let const_of_user_written_attribute ?legacy_immediate ~context
+    Location.{ loc; txt = attribute } =
+  let unchecked = Const.of_user_written_attribute_unchecked attribute in
+  let checked =
+    check_user_written_annotation ?legacy_immediate ~context ~loc unchecked
+  in
+  Location.{ loc; txt = checked }
+
+let const_of_attributes ~legacy_immediate ~context attrs =
+  Builtin_attributes.jkind ~legacy_immediate attrs
+  |> Result.map
+       (Option.map (const_of_user_written_attribute ~legacy_immediate ~context))
+
+let of_annotated_const ~context Location.{ txt = const; loc = const_loc } =
+  of_const ~why:(Annotated (context, const_loc)) const
 
-let of_annotation_option ?legacy_immediate ~context =
-  Option.map (of_annotation ?legacy_immediate ~context)
+let of_annotated_const_option ~context =
+  Option.map (of_annotated_const ~context)
+
+let of_annotated_const_default ~context ~default =
+  Option.fold ~none:default ~some:(of_annotated_const ~context)
+
+let of_annotation ?legacy_immediate ~context (annot : _ Location.loc) =
+  let const =
+    const_of_user_written_annotation ?legacy_immediate ~context annot
+  in
+  let jkind = of_annotated_const { txt = const; loc = annot.loc } ~context in
```

Comment:
This code looks pretty similar in the stacked PR.

Are you suggesting passing two arguments to `of_annotated_const` rather than the product of const and location? That sounds good to me. (Let me know and I'll make that change.)

Reply:
Yes, I think I am. Pretty minor change, really, but I suppose it saves an allocation.

Reply:
Done in the stacked PR, where it is indeed easier.